modules/040-node-manager/images/fencing-agent/internal/agent/controller.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2024 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package agent
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"fecning-controller/internal/watchdog"
+
+	"go.uber.org/zap"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	FecningNodeLabel = "node-manager.deckhouse.io/fencing-enabled"
+)
+
+var maintanenceAnnotations = [...]string{
+	`update.node.deckhouse.io/disruption-approved`,
+	`update.node.deckhouse.io/approved`,
+	`node-manager.deckhouse.io/fencing-disable`,
+}
+
+type FencingAgent struct {
+	logger     *zap.Logger
+	config     Config
+	kubeClient kubernetes.Interface
+	watchDog   watchdog.WatchDog
+}
+
+func NewFencingAgent(logger *zap.Logger, config Config, kubeClient kubernetes.Interface, wd watchdog.WatchDog) *FencingAgent {
+	l := logger.With(zap.String("node", config.NodeName))
+	return &FencingAgent{
+		logger:     l,
+		config:     config,
+		kubeClient: kubeClient,
+		watchDog:   wd,
+	}
+}
+
+func (fa *FencingAgent) setNodeLabel(ctx context.Context) error {
+	node, err := fa.kubeClient.CoreV1().Nodes().Get(ctx, fa.config.NodeName, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	node.Labels[FecningNodeLabel] = ""
+	_, err = fa.kubeClient.CoreV1().Nodes().Update(ctx, node, v1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (fa *FencingAgent) removeNodeLabel() error {
+	node, err := fa.kubeClient.CoreV1().Nodes().Get(context.TODO(), fa.config.NodeName, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	delete(node.Labels, FecningNodeLabel)
+	_, err = fa.kubeClient.CoreV1().Nodes().Update(context.TODO(), node, v1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (fa *FencingAgent) startWatchdog(ctx context.Context) error {
+	var err error
+	fa.logger.Info("Arm the watchdog")
+	err = fa.watchDog.Start()
+	if err != nil {
+		return err
+	}
+	fa.logger.Info("Set fencing node label", zap.String("label", FecningNodeLabel))
+	err = fa.setNodeLabel(ctx)
+	if err != nil {
+		// We must stop watchdog if we can't set nodelabel
+		fa.logger.Error("Unable to set node label, so disarming watchdog...")
+		_ = fa.watchDog.Stop()
+		return err
+	}
+	return nil
+}
+
+func (fa *FencingAgent) startLiveness() {
```

Comment:
i think we need context here and stop server if context is cancelled

Reply:
fixed